site2/docs/schema-understand.md

```diff
@@ -0,0 +1,321 @@
+            .topic(TOPIC)
+            ...
+            .subscribe();
+
+        // receive key/value pair
+        Message<KeyValue<Integer, String>> msg = consumer.receive();
+        KeyValue<Integer, String> kv = msg.getValue();
+        ```
+
+* **Complex type 2: `struct`**
+
+    Pulsar uses [Avro Specification](http://avro.apache.org/docs/current/spec.html) to declare the schema definition for `struct` schema. This allows Pulsar:
+
+    * to use same tools to manage schema definitions
+
+    * to use different serialization/deserialization methods to handle data
+
+    There are two methods to use `struct` schema. 
+
+  * **Method 1: static**
+
+      You can predefine the `struct` schema, and it can be a POJO in Java, a `struct` in Go, or classes generated by Avro or Protobuf tools. 
+
+      **Example** 
+
+      Pulsar gets the schema definition from the predefined `struct` using an Avro library. The schema definition is the schema data stored as a part of the schema info.
+
+      1. Create the _User_ class to define the messages sent to Pulsar topics.
+
+          ```text
+          public class User {
+              String name;
+              int age;
+          }
+          ```
+
+      2. Create a producer with a `struct` schema and send messages.
+
+          ```text
+          Producer<User> producer = client.newProducer(Schema.AVRO(User.class)).create();
+          producer.newMessage().value(User.builder().userName("pulsar-user").userId(1L).build()).send();
+          ```
+
+      3. Create a consumer with a `struct` schema and receive messages
+
+          ```text
+          Consumer<User> consumer = client.newConsumer(Schema.AVRO(User.class)).create();
+          User user = consumer.receive();
+          ```
+
+  * **Method 2: generic**
+
+      Sometimes applications do not have pre-defined structs, and you can use this method to define schema and access data.
+
+      You can define the `struct` schema using the `GenericSchemaBuilder`, generate a generic struct using `GenericRecordBuilder` and consume messages into `GenericRecord`.
+
+      **Example** 
+
+      1. Use `RecordSchemaBuilder` to build a schema.
+
+          ```text
+          RecordSchemaBuilder recordSchemaBuilder = SchemaBuilder.record("schemaName");
+          recordSchemaBuilder.field("intField").type(SchemaType.INT32);
+          SchemaInfo schemaInfo = recordSchemaBuilder.build(SchemaType.AVRO);
+
+          Producer<GenericRecord> producer = client.newProducer(Schema.generic(schemaInfo)).create();
+          ```
+
+      2. Use `RecordBuilder` to build the struct records.
+
+          ```text
+          producer.newMessage().value(schema.newRecordBuilder()
+                      .set("intField", 32)
+                      .build()).send();
+          ```
+
+### Auto Schema
+
+If you do not know the schema type of a producer, the schema type of a Pulsar topic, or the schema type a consumer, you might have incompatible issues. 
```

Comment:
@sijie I add this based on my understanding, is it correct?

Reply:
If you don't know the schema type of a Pulsar topic ahead of time, you can use AUTO schema to produce or consume generic records to or from brokers.